### ISSUE TITLE
Multiple calls to same snippet in page do not respect uncached directive.

### ISSUE THREAD

**Issue by ByteJamRob** [modxcms/revolution#11338]
Tested in Revo 2.2.14 and 2.2.6.

To test, make a snippet called TEST with the following code:

return md5(uniqid(rand(),1));

Drop the following snippet calls on a page:

<p>[[!TEST]]</p>

<p>[[!TEST]]</p>

<p>[[!TEST]]</p>

Load the page. You will see the same value repeated three times. Since I have told the snippet to be uncached, this is not what I would expect to see.

A workaround is to assign a dummy parameter with different values for each snippet call:

<p>[[!TEST? &dummy=`1`]]</p>

<p>[[!TEST? &dummy=`2`]]</p>

<p>[[!TEST? &dummy=`3`]]</p>

Even though the "dummy" parameter is not referenced in the snippet's source code, this will yield the desired results.

Can Revo be updated so that these two scenarios behave the same way?

**Comment by whitebyte** [comment]
MODX docs persistently treats snippets like "functions". If we pass different params to a function, we expect different results. Given that we call snipped uncached AND with different params it's pretty weird to have cached results. I suggest to treat this issue as a bug, not as a feature. It can affect parsing speed, yet this should be handled.

**Comment by Bruno17** [comment]
Don't you get different outputs with different params?

**Comment by HitkoDev** [comment]
@OptimusCrime It doesn't use the footprint to optimize speed, the snippet is still executed on each call, and the value from the LAST execution is inserted for each snippet tag (see https://github.com/modxcms/revolution/issues/12715)

**Comment by inreti** [comment]
I have run into this years before, where a snippet had to output timing informations...

If you see the MODX request and parsing from a functional programming point of view, everything said about same footprint = same result can be considered as right. But this somehow interferes with the concept of caching. The first computes the same value everytime, a cached value just returns a previously computed value. 

[[!snippet]] should be computed every time it occurs imho. That's would people understand as "uncached" and what they expect then. Me too. So if you dare to use 10 times a snippet call like that on a page, fine, you know what you do and what performance issues result out of them. You would expect that performance thing then, so for me that's no problem here. 

If actually the uncached snippets are evaluated each time, but only the last value makes it to the result and all calls are replaced by that result, then it is a bug - depending on the concept the bug is either to evaluate the snippet more than once _or_ the bug is using only the last result for all calls.

**Comment by HitkoDev** [comment]
@inreti You can demonstrate this by setting a global variable and increasing it each time the snippet is evaluated

**Comment by sottwell** [comment]
The problem is in the properties. They have the same name, so each one overwrites the previous one. Perhaps a mechanism for "snippetID" and use that as a prefix for the property, such as &snippetID=1 &property1=`blah` and for the next one &snippedID=2 &property2=`foo` etc. then "proptery1" would be "property" for snippet1, and "property2" would be "property" for snippet2.

**Comment by HitkoDev** [comment]
MODX parses all snippets first, presumably building an array {footprint => output}, then renders the output into the content by replacing footprint1 with output1, footprint2 with output2, etc. Although the snippet is evaluated each time, since it has the same tag footprint the last output overrides any previous output.

**Comment by inreti** [comment]
Hmmm. So maybe we can just add a virtual param (random hash) to a snippets signature if it is called uncached? That would make the footprints unique...

**Comment by HitkoDev** [comment]
Fixed it by replacing an associative array with two separate arrays: https://github.com/modxcms/revolution/pull/12716

**Comment by OptimusCrime** [comment]
Like I said, this is confirmed the intended behavior by @opengeek 

Questions to avoid this behavior can be asked in the forums or at Slack.

As it is not a bug I'll close this.

@modxbot close

**Comment by Mark-H** [comment]
@OptimusCrime where did you see Jason comment on this issue?

**Comment by jpdevries** [comment]
This is closest thing I found on RTFM.

> Revolution, any tag can be called uncached by inserting an exclamation point immediately after the double-bracket: [[!snippet]], [[!$chunk]], [[!+placeholder]], [[!*template_var]], etc.
&emsp;&mdash;[Tag Syntax](https://rtfm.modx.com/revolution/2.x/making-sites-with-modx/tag-syntax)

Please note it says *any tag* not any tag the first time it is included on a page. If this is the intended behavior it seems like RTFM needs to be updated.

**Comment by HitkoDev** [comment]
Let's recap:
- unacahed snippets with identical footprint cause collisions while parsing
- they don't behave like cached snippets either because the output comes from the last call not from the first one
- all of them are parsed despite only the last output is actually inserted
- hacks are needed to achieve the "intuitive" result
- there's no situation where user would want to execute the same snippet multiple times and display the final result each time; it's much more efficient to execute the code multiple times within the snippet itself and then output it
Yet @OptimusCrime still claims it's not a bug...

**Comment by OptimusCrime** [comment]
@Mark-H I've talked to him about this before. There are multiple sources for this behavior.

**Comment by OptimusCrime** [comment]
All blocks with identical footprints will run once and produce one output only. The only difference between an uncached and a cached block is that a cached is only ran once. The behavior for their output is the same (run once).

**Comment by Mark-H** [comment]
@OptimusCrime According to the steps and code provided in #12715, each call is actually _parsed individually_ - but their _output_ is overridden by the last one. 

That's different from what you're describing where a second tag (that is the same as the first tag) is not parsed at all.

**Comment by OptimusCrime** [comment]
Hmm, the way I see it, this is not a duplicate of #12715.

I am trying to figure out where I and opengeek discussed this but I have not had any luck yet.

I could reopen the issue if you want to? Or @opengeek could come in here and close this argument once and for all.

**Comment by Mark-H** [comment]
I do think this issue and #12715 are the same (at least from a code standpoint), they're just explaining them from a different perspective. Right now I believe based on the discussions here the expected behaviour should be either one of these two:

1. Tags with the same footprint need to be executed only once, and the output of each is the same.
2. Tags are parsed individually, and the output of each can be different. 

While it sounds like 1) would be the current behaviour (and I actually thought it was, too), #12715 proves it isn't: tags with the same footprint are parsed individually, but their output of each is the same. That's a combination of 1) and 2) that doesn't make much sense. Why bother parsing it if you're going to discard the output?

I'm on board with fixing this behaviour in the next 2.x release. Preferably by adjusting the behaviour to 2), as adjusting the behaviour to 1) sounds like it could cause unintended changes if all of a sudden snippets are no longer executed while they were before.

**Comment by HitkoDev** [comment]
https://github.com/modxcms/revolution/pull/12719 changes the behaviour to 2).

**Comment by opengeek** [comment]
@Mark-H my concern was that adjusting parsing to do 2) would impact parsing and performance significantly. Consider placeholders used multiple times in a chunk—instead of one processing and simple `str_replace()` call, you will now process it for each occurrence.

Keep this open. I'm open to making a change as proposed in #12719, but we should discuss if this is or is not a BC break. In other words, we need to determine whether this change must be included in `3.x` instead of next minor (via `2.x`).

**Comment by OptimusCrime** [comment]
@modxbot open

**Comment by Bruno17** [comment]
maybe a tag-property ```&forceparsing=`1` ``` or something?

**Comment by HitkoDev** [comment]
I understand one could assume same footprint == same output to improve parsing speed. In that case a 'strictly uncached' token should be introduced to defer that behaviour when needed. However, looking at previous versions of modParser I can see Revo 2 has allways parsed each tag separately.

**Comment by VoiD2008** [comment]
simple workaround use additional different param.
snippet [[rnd]]:
`<?php return rand(1,100);`
chunk:
`[[rnd]] / [[rnd]]`
will return same integers

but
chunk:
`[[rnd? &seed=1]] / [[rnd? &seed=2]]`
will return different (or same in random =)) )

**Comment by sepiariver** [comment]
Interesting thread. Intuitively I would think any uncached element would be parsed individually at runtime. An uncached element might use some uncached input such as that from a user, and could have different output despite having the same signature.

**Comment by opengeek** [comment]
@sepiariver Doesn't work that way. If the parser sees a tag signature that it has already parsed in a single request, it simply uses the content returned the first time it was parsed on every instance it encounters.

**Comment by opengeek** [comment]
And user input would be the same within the same request. You can't have two different user inputs to the same snippet on the same request.

**Comment by sepiariver** [comment]
Thanks for the clarification :) Understand how it's working currently. Still of the opinion that intuitively an "un-cacheable" element would be processed individually. Inputs can, and sometimes (maybe rarely?) do, change over the course of a request. User input (or any variable) can be modified by another Snippet, etc.

This issue hasn't caused me much problems, personally. And you're right in that the parser is more performant the way it is. Only remarking that "intuitively" when I think of an "un-cacheable" element, I think it's going to be processed individually on every request.

**Comment by Ibochkarev** [comment]
### Triage (current 3.x)

Still the same parser behavior. @opengeek clarified it is intentional: within one request, identical tag signatures reuse the first result (tag/`elementCache` by signature). `[[!…]]` means “don’t put this in the resource cache,” not “re-execute every identical call on the page.”

Workaround remains distinct properties (`&seed=` / `&dummy=`). Disposition: close as wontfix / document, or only change if the team wants a new explicit opt-in for per-instance uncached execution.
